File: MCP_Server/tools/editor_tools.py
from mcp.server.fastmcp import FastMCP
from client.unreal_client import UnrealClient


def register_editor_tools(mcp: FastMCP, client: UnrealClient):

    @mcp.tool()
    async def run_console_command(command: str) -> str:
        result = client.send_command("run_console_command", {"command": command})
        if result.get("success"):
            return f"Command executed: {result.get('result', 'OK')}"
        return f"Command failed: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def save_current_level() -> str:
        result = client.send_command("save_current_level", {})
        if result.get("success"):
            return "Level saved successfully"
        return f"Failed to save level: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def play_in_editor() -> str:
        result = client.send_command("play_in_editor", {})
        if result.get("success"):
            return "Started Play In Editor"
        return f"Failed: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def stop_play_in_editor() -> str:
        result = client.send_command("stop_play_in_editor", {})
        if result.get("success"):
            return "Stopped Play In Editor"
        return f"Failed: {result.get('error', 'Unknown error')}"

    @mcp.tool()
    async def get_editor_info() -> str:
        result = client.send_command("get_editor_info", {})
        if result.get("success"):
            info = result.get("result", {})
            return f"Unreal Editor Info: {info}"
        return f"Failed: {result.get('error', 'Unknown error')}"

File: MCP_Server/src/tools/editor_tools.rs
use rmcp::model::*;
use rmcp::schemars;
use rmcp::server::tool;
use serde_json::json;
use crate::unreal_client::UnrealClient;
use std::sync::Arc;
use tokio::sync::Mutex;

pub struct EditorTools {
    client: Arc<Mutex<UnrealClient>>,
}

impl EditorTools {
    pub fn new(client: Arc<Mutex<UnrealClient>>) -> Self {
        Self { client }
    }

    #[tool(description = "Run a console command in Unreal Engine")]
    async fn run_console_command(
        &self,
        #[tool(param)]
        #[schemars(description = "Console command string")]
        command: String,
    ) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("run_console_command", json!({"command": command})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    "Command executed successfully".to_string()
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Save the current level")]
    async fn save_current_level(&self) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("save_current_level", json!({})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    "Level saved successfully".to_string()
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Start Play In Editor (PIE)")]
    async fn play_in_editor(&self) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("play_in_editor", json!({})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    "Started Play In Editor".to_string()
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Stop Play In Editor (PIE)")]
    async fn stop_play_in_editor(&self) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("stop_play_in_editor", json!({})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    "Stopped Play In Editor".to_string()
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }

    #[tool(description = "Get Unreal Editor information")]
    async fn get_editor_info(&self) -> Result<CallToolResult, McpError> {
        let mut client = self.client.lock().await;
        match client.send_command("get_editor_info", json!({})).await {
            Ok(response) => {
                let text = if response["success"].as_bool().unwrap_or(false) {
                    format!("Editor Info: {}", response["result"])
                } else {
                    format!("Failed: {}", response["error"])
                };
                Ok(CallToolResult::success(vec![Content::text(text)]))
            }
            Err(e) => Ok(CallToolResult::success(vec![Content::text(format!("Error: {}", e))])),
        }
    }
}
